service/src/PlatformTopo.cpp

```diff
@@ -418,30 +460,79 @@ namespace geopm
         PlatformTopoImp::create_cache();
     }
 
+    void PlatformTopo::create_cache_service(void)
+    {
+        PlatformTopoImp::create_cache_service();
+    }
+
     void PlatformTopoImp::create_cache(void)
     {
         PlatformTopoImp::create_cache(M_CACHE_FILE_NAME);
     }
 
+    void PlatformTopoImp::create_cache_service(void)
+    {
+        PlatformTopoImp::create_cache(M_SERVICE_CACHE_FILE_NAME);
+    }
+
     void PlatformTopoImp::create_cache(const std::string &cache_file_name)
     {
-        // If cache file is not present, create it
-        struct stat cache_stat;
-        if (stat(cache_file_name.c_str(), &cache_stat)) {
-            std::string cmd = "out=" + cache_file_name + ";"
-                              "lscpu -x > $out && "
-                              "chmod a+rw $out";
+        // If cache file is not present, or is too old, create it
+        bool is_file_ok = false;
+        try {
+            is_file_ok = check_file(cache_file_name);
+        }
+        catch (const geopm::Exception &ex) {
+            // sysinfo or stat failed; file does not exist (2) or permission denied (13)
+            if (ex.err_value() == EACCES) {
+                throw; // Permission was denied; Cannot create files at the desired path
+            }
+        }
+
+        if (is_file_ok == false) {
+            mode_t perms;
+            if (cache_file_name == M_SERVICE_CACHE_FILE_NAME) {
+                perms = S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH; // 0o644
+            }
+            else {
+                perms = S_IRUSR | S_IWUSR; // 0o600
+            }
+
+            std::string tmp_string = cache_file_name + "XXXXXX";
+            char tmp_path[NAME_MAX];
+            strncpy(tmp_path, tmp_string.c_str(), NAME_MAX);
+            int tmp_fd = mkstemp(tmp_path);
+            if (tmp_fd == -1) {
+                throw Exception("PlatformTopo::create_cache(): Could not create temp file: ",
+                                errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+            }
+            int err = fchmod(tmp_fd, perms);
+            if (err) {
```

Comment:
In case of error, before throwing: `close(tmp_fd)`

However, seeing the intention is to close it, anyway, you could simply close it right after the call to `fchmod()` regardless of return value.